### ISSUE TITLE
NotImplementedError: can not rewrite Univariate Quotient Polynomial Ring ..

### ISSUE THREAD

**Issue by swewers** [MCLF/mclf#105]
The following gives an error:

    sage: F.<x> = FunctionField(QQ)
    sage: R.<y> = F[]
    sage: G = y^4 - 4*x^4 + 48*x^3 - 24*x^2 + 48*x - 132
    sage: v_2 = QQ.valuation(2)
    sage: v0 = GaussValuation(F._ring, v_2).augmentation(F._ring.gen()+1, 5/4)
    sage: v = F.valuation(v0)
    sage: v.mac_lane_approximants(G)

    NotImplementedError: can not rewrite Univariate Quotient Polynomial Ring in u2 over Fraction Field of    
    Univariate Polynomial Ring in x over Finite Field of size 2 (using GF2X) with modulus u2^2 + 1/x^3 as 
    an isomorphic ring
    sage:

**Comment by saraedum** [comment]
I see. The code only handles number fields and finite fields but no function fields yet. This should be fixed upstream and  then monkey-patched here.

**Comment by swewers** [comment]
But it usually works! There must be something specific about this example which makes it fail. 
Isn't the MacLane algorithm completely generic for discrete valuations (provided you have functions like ``reduce``, ``lift`` or ``element_with_valuation``)?

**Comment by saraedum** [comment]
True, but for some reason the code wants to replace the fraction field with a function field to improve performance. But that doesn't work yet.

**Comment by saraedum** [comment]
What happens is that a factorization needs to be computed and we don't know how to factor in a quotient so we try to rewrite it to a field with more features.

**Comment by saraedum** [comment]
Changing that does not solve the problem as the residue field is an extension of a rational function field and we don't know how to factor there…this is now: https://trac.sagemath.org/ticket/26123, https://trac.sagemath.org/ticket/26124.

**Comment by saraedum** [comment]
#16564 is currently waiting to be reviewed :)

**Comment by JRSijsling** [comment]
To try to review that ticket, I checked out that ticket and rebuilt, at first with `sage -b` and later using `make`. The doctests all fail and even defining a function field fails. This failure likely has nothing to do with your ticket, but I am at a loss what to do and whether this is a real bug that I should report (and if so, where) or just me being dense.

Gist [here](https://gist.github.com/JRSijsling/2cbbc3fd12ead3fac9840d76a07c7dfd).

**Comment by saraedum** [comment]
Thanks for having a look. Seems like something broke when merging this with the latest beta.

Having to rebuild is quite annoying. I hope we could finally get https://trac.sagemath.org/ticket/24842 in so people would just have to click a button to be able to review a specific build of Sage…

**Comment by JRSijsling** [comment]
Either that, or there is a problem with the last beta itself. I do not have that built right now, but I would not exclude the possibility. The diff is very small.

**Comment by JRSijsling** [comment]
In fact no, the develop branch is all right. It seems that `FunctionFieldVectorSpaceIsomorphism` got replaced by a function with identical purposes named `FunctionFieldIsomorphism`. Will modify and rebuild to see what happens.

**Comment by JRSijsling** [comment]
This works, though there are still some minor issues with the doctests. I will mention this on trac, where it belongs.

**Comment by saraedum** [comment]
`make build` builds everything but the documentation. Is that what
you've been looking for?

**Comment by JRSijsling** [comment]
Yes that is exactly it, thank you!

**Comment by swewers** [comment]
I don't understand what the problem with this bloody patchbot is! What can one do in addition to writing a positive review?
